explorer/interpreter/type_checker.cpp

```diff
@@ -1359,75 +1374,135 @@ class TypeChecker::ConstraintTypeBuilder {
-    // TODO: Avoid the quadratic behavior.
+  // Check for conflicting rewrites and deduplicate.
+  auto DeduplicateRewrites(SourceLocation source_loc) -> ErrorOr<Success> {
     std::vector<RewriteConstraint> new_rewrite_constraints;
     for (auto& rewrite_a : rewrite_constraints_) {
-      bool found_existing = false;
-      for (auto& rewrite_b : new_rewrite_constraints) {
-        if (ValueEqual(rewrite_a.constant, rewrite_b.constant, std::nullopt)) {
-          if (ValueEqual(rewrite_a.unconverted_replacement,
-                         rewrite_b.unconverted_replacement, std::nullopt) &&
-              TypeEqual(rewrite_a.unconverted_replacement_type,
-                        rewrite_b.unconverted_replacement_type, std::nullopt)) {
-            found_existing = true;
-            break;
-          }
-          return ProgramError(source_loc)
-                 << "multiple different rewrites for `" << *rewrite_a.constant
-                 << "`:\n"
-                 << "  " << *rewrite_b.unconverted_replacement << "\n"
-                 << "  " << *rewrite_a.unconverted_replacement;
+      if (auto existing_rewrite = LookupRewrite(
+              new_rewrite_constraints, &rewrite_a.constant->interface(),
+              &rewrite_a.constant->constant())) {
+        auto& rewrite_b = **existing_rewrite;
+        if (ValueEqual(rewrite_a.unconverted_replacement,
+                       rewrite_b.unconverted_replacement, std::nullopt) &&
+            TypeEqual(rewrite_a.unconverted_replacement_type,
+                      rewrite_b.unconverted_replacement_type, std::nullopt)) {
+          // This is a duplicate, ignore it.
+          continue;
         }
+        return ProgramError(source_loc)
+               << "multiple different rewrites for `" << *rewrite_a.constant
+               << "`:\n"
+               << "  " << *rewrite_b.unconverted_replacement << "\n"
+               << "  " << *rewrite_a.unconverted_replacement;
       }
-      if (!found_existing) {
-        new_rewrite_constraints.push_back(rewrite_a);
-      }
+      new_rewrite_constraints.push_back(rewrite_a);
     }
     rewrite_constraints_ = std::move(new_rewrite_constraints);
+    return Success();
+  }
 
-    // Rebuild a value with our updated rewrite constraints. Because this
-    // builder is in scope as a partial constraint type, if we perform a null
-    // substitution any nested associated constants will be rebuilt.
-    auto rebuild = [&](Nonnull<const Value*> value) {
-      Bindings null_bindings;
-      return type_checker.SubstituteImpl(null_bindings, value);
-    };
+  // Apply rewrites to each other and find a fixed point, or diagnose if there
+  // is a cycle.
+  auto ApplyRewritesToRewrites(TypeChecker& type_checker,
+                               SourceLocation source_loc) -> ErrorOr<Success> {
+    // Add this builder to the type checker's scope so that it considers our
+    // rewrites.
+    type_checker.partial_constraint_types_.push_back(this);
```

Comment:
Are you expecting this to be called from more places that you didn't want to do this from the caller, given that ApplyRewritesToConstraints does the same?

Reply:
It might be, but my main concern is that this is necessary for the function to have a meaningful contract.